### PR TITLE
Textbook listing

### PR DESCRIPTION
**Pull request by emumoo**
@aw3as yo

### CODE REVIEW

File: app/controllers/textbook_transactions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
+class TextbookTransactionsController < ApplicationController
+
+  def index
+    @textbook_transactions = TextbookTransaction.active.map do |tt|
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: pls better name than tt something like transaction ty

File: app/controllers/textbook_transactions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    @books = Book.order("RAND()").map do |book|
+      {
+        :image => (book.small_image_link ? book.small_image_link : "/assets/icons/no_book.png"),
+        :title => book.title,
+        :id => book.id
+      }
+    end
+  end
+
+  def get_book_titles
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: consolidate this method with prior one, use respond_to do |format|

File: app/controllers/textbook_transactions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        :title => book.title,
+        :image => (book.small_image_link ? book.small_image_link : "/assets/icons/no_book.png")
+      }
+    end
+
+    render :json => [results]
+  end
+
+  def claim
+    puts params
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: no puts here

File: app/controllers/textbook_transactions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    cell = params[:cellphone].strip
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: should probably do better than strip, to handle dashes, hyphens, parentheses

File: app/controllers/textbook_transactions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    current_user.update(:cellphone => cell)
+    transaction = TextbookTransaction.active.find(params[:claim_id])
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: why search through active transactions? incur additional sql queries

File: app/controllers/textbook_transactions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+    if cell.length == 10
+      response = RestClient.post 'http://textbelt.com/text', :number => transaction.seller.cellphone, :message => "Your posting for \"#{transaction.book.title}\" has been claimed!\nContact info: #{current_user.cellphone}"
+      if JSON.parse(response)["success"]
+        puts 'hihihihihihihihihihi'
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: heyheyheyheyheyhey taketaketaketaketake outoutoutoutoutout putsputsputsputsputsputs

Comment (emumoo) [inline app/controllers/textbook_transactions_controller.rb]: okokokokokokokok

File: app/controllers/textbook_transactions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        transaction.update(:buyer_id => current_user.id)
+        transaction.update(:sold_at => Time.now)
+        render :json => {
+          status: "success"
+        }
+      else 
+        render :json => {
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: http://stackoverflow.com/questions/674027/what-should-a-json-service-return-on-failure-error format your json responses like so, and give a http error code

File: app/controllers/textbook_transactions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      render :json => {
+        status: "Badly formatted phone number"
+      }
+    end
+  end
+
+  def create
+    current_user.update(:cellphone => params[:cellphone])
+
+    if params[:cellphone].length == 10
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: you should move these kind of validations to javascript, to increase responsiveness

File: app/controllers/textbook_transactions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+  def update
+  end
+
+  def destroy
+  end
+
+  private
+    # Never trust parameters from the scary internet, only allow the white list through.
+    def textbook_transaction_params
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: http://api.rubyonrails.org/classes/ActionController/ParamsWrapper.html pls

File: app/controllers/textbook_transactions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+      end
+    else
+      render :json => {
+        status: "Badly formatted phone number"
+      }
+    end
+    
+  end
+
+  def new
```
Comment (aw3as) [inline app/controllers/textbook_transactions_controller.rb]: do you use any of these methods? delete if you don't

File: app/helpers/textbook_transactions_helper.rb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+module TextbookTransactionsHelper
```
Comment (aw3as) [inline app/helpers/textbook_transactions_helper.rb]: no

File: app/models/textbook_transaction.rb [hunk 12]
```diff
@@ -0,0 +1,14 @@
+class TextbookTransaction < ActiveRecord::Base
+	default_scope { order('created_at DESC') }
```
Comment (aw3as) [inline app/models/textbook_transaction.rb]: no default scope - we talked about this

File: app/models/textbook_transaction.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+	belongs_to :seller, :class_name => 'User', :foreign_key => 'seller_id'
```
Comment (aw3as) [inline app/models/textbook_transaction.rb]: not sure if you need :foreign_key it should know it based on belongs_to :seller - try it to see if works

File: app/models/user.rb [hunk 14]
```diff
@@ -14,6 +14,13 @@ class User < ActiveRecord::Base
   has_and_belongs_to_many :courses
   has_many :schedules, dependent: :destroy
 
+  # relationships for textbook listings
+  has_many :textbook_listings, :class_name => 'TextbookTransaction', :foreign_key => 'seller_id'
```
Comment (aw3as) [inline app/models/user.rb]: you don't need string, just :class_name => TextbookTransaction without quotes work fine

File: app/views/layouts/application.html.slim [hunk 15]
```diff
@@ -54,6 +54,9 @@ html
       -# Necessary for rendering scheduler modals correctly      
       - if controller.controller_name == "scheduler" && controller.action_name == "scheduler"
         = render 'scheduler/modals'
+      -# Necessary for rendering textbook_transaction modals correctly
+      - if controller.controller_name == "textbook_transactions" ||controller.controller_name == "books"
```
Comment (aw3as) [inline app/views/layouts/application.html.slim]: pls put space after ||

File: config/environments/development.rb [hunk 16]
```diff
@@ -31,4 +31,7 @@
   config.action_mailer.delivery_method = :smtp
   #change to correct email settings
   config.action_mailer.smtp_settings = {:address => "localhost", :port => 1025}
+
+  # Ignore pending migrations
+  config.active_record.migration_error = false#:page_load
```
Comment (aw3as) [inline config/environments/development.rb]: fu leave this in

File: config/routes.rb [hunk 17]
```diff
@@ -42,15 +43,27 @@
 
   resources :subdepartments, :only => [:show]
 
-  resources :books, :only => [:index]
+  resources :books, :only => [:index, :show]
```
Comment (aw3as) [inline config/routes.rb]: you should move the get below this to above this line - for good juju

File: config/routes.rb [hunk 18]
```diff
@@ -42,15 +43,27 @@
...
     collection do
       get :search_subdepartment
     end
   end
  
+  resources :textbook_transactions, :only => [:index, :create]
+  # Autocomplete for textbook titles
+  resources :textbook_transactions, :only => [] do
+    collection do
+      get :get_book_titles
```
Comment (aw3as) [inline config/routes.rb]: NO

File: spec/controllers/textbook_transactions_controller_spec.rb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe TextbookTransactionsController, :type => :controller do
```
Comment (aw3as) [inline spec/controllers/textbook_transactions_controller_spec.rb]: do you use this

File: spec/factories/textbook_transactions.rb [hunk 20]
```diff
@@ -0,0 +1,6 @@
+FactoryGirl.define do
```
Comment (aw3as) [inline spec/factories/textbook_transactions.rb]: no commit specs

File: spec/helpers/textbook_transactions_helper_spec.rb [hunk 21]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
```
Comment (aw3as) [inline spec/helpers/textbook_transactions_helper_spec.rb]: no helper either

File: textbook_transactions/textbook_exchange.sql [hunk 22]
```diff
@@ -0,0 +1,24 @@
+# Author Lawrence Hook
+
+CREATE TABLE `textbook_transactions` (
+	`id` int(11) NOT NULL AUTO_INCREMENT,
+	`seller_id` int(11) DEFAULT NULL,
```
Comment (aw3as) [inline textbook_transactions/textbook_exchange.sql]: not null bbz

File: textbook_transactions/textbook_exchange.sql [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+	`buyer_id` int(11) DEFAULT NULL,
+	`book_id` int(11) NOT NULL,
+	`price` int(11) NOT NULL,
+	`condition` varchar(255) NOT NULL,
+	`notes` text DEFAULT NULL,
+	`created_at` datetime NOT NUll,
+	`updated_at` datetime NOT NUll,
+	`sold_at` datetime DEFAULT NULL,
+	PRIMARY KEY (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```
Comment (aw3as) [inline textbook_transactions/textbook_exchange.sql]: add some indices here on book_id, seller_id, maybe buyer_id

**Pull request closed without merging**